physics/eclipse_test.cpp

```diff
@@ -0,0 +1,167 @@
+using quantities::si::Minute;
+using quantities::si::Metre;
+using quantities::si::Nano;
+using quantities::si::Milli;
+using quantities::si::Kilo;
+using quantities::ArcCos;
+using testing_utilities::AbsoluteError;
+using ::testing::Gt;
+using ::testing::Lt;
+using ::testing::AllOf;
+
+namespace physics {
+
+class EclipseTest : public testing::Test {
+ protected:
+  EclipseTest() {
+    solar_system_1950_.Initialize(
+        SOLUTION_DIR / "astronomy" / "gravity_model.proto.txt",
+        SOLUTION_DIR / "astronomy" /
+            "initial_state_jd_2433282_500000000.proto.txt");
+  }
+
+  SolarSystem<ICRFJ2000Equator> solar_system_1950_;
+};
+
+TEST_F(EclipseTest, Dummy) {
+  auto ephemeris = solar_system_1950_.MakeEphemeris(
+      McLachlanAtela1992Order5Optimal<Position<ICRFJ2000Equator>>(),
+      45 * Minute, 5 * Milli(Metre));
+
+  ephemeris->Prolong(JulianDate(2433374.5));  // Prolong until date of eclipse
+                                              // (Eclipse was 1950-04-02 but JD
+                                              // is 1950-04-03:00:00:00)
+  // pass body to Ephemeris.trajectory
+  auto sun = ephemeris->bodies()[solar_system_1950_.index("Sun")];
+  auto earth = ephemeris->bodies()[solar_system_1950_.index("Earth")];
+  auto moon = ephemeris->bodies()[solar_system_1950_.index("Moon")];
+
+  // Massive_body eventually needs radius information. Or non-hardcoded data
+  // pulled from
+  // https://github.com/mockingbirdnest/Principia/blob/master/astronomy/gravity_model.proto.txt
+  auto r_sun = 696000.0 * Kilo(Metre);
+  auto r_earth = 6378.1363 * Kilo(Metre);
+  auto r_moon = 1738.0 * Kilo(Metre);
+
+  // Dates are TDB Julian Day for 1948-04-02
+  auto P1 = JulianDate(2433374.25788409);  // 18:10:49 UT
+  auto U1 = JulianDate(2433374.29850909);  // 19:09:19
+  auto U2 = JulianDate(2433374.354979);    // 20:30:38
+  auto U3 = JulianDate(2433374.37367113);  // 20:57:33
+  auto U4 = JulianDate(2433374.43016419);  // 22:18:54
+  auto P4 = JulianDate(2433374.47075446);  // 23:17:21
+
+  // Lunar eclipse -- U1 (time is start of umbral phase)
+  auto q_sun = ephemeris->trajectory(sun)->EvaluatePosition(U1, nullptr);
+  auto q_moon = ephemeris->trajectory(moon)->EvaluatePosition(U1, nullptr);
+  auto q_earth = ephemeris->trajectory(earth)->EvaluatePosition(U1, nullptr);
+
+  // check body angles at target time
+  // Earth/Sun lineup
+  auto alpha = ArcSin((r_sun - r_earth) / (q_sun - q_earth).Norm());
+  auto q_U14 =
+      q_earth + Normalize(q_earth - q_sun) * (r_earth + r_moon) / Sin(alpha);
+  // Earth/Moon lineup for U14
+  auto gamma = ArcCos(InnerProduct(q_U14 - q_earth, q_U14 - q_moon) /
+                      ((q_U14 - q_moon).Norm() * (q_U14 - q_earth).Norm()));
+  // Do the angles work?
+  LOG(ERROR) << alpha;
+  LOG(ERROR) << gamma;
+  EXPECT_THAT(AbsoluteError(alpha, gamma),
+              AllOf(Lt(1.0 * Milli(Radian)), Gt(1.0 * Nano(Radian))));
+
+  // Lunar eclipse -- U2 (time is start of full eclipse)
```

Comment:
Hmm.  By now the alert reader notices a pattern: it seems that you are doing the exact same computation four times for `U1` to `U4` (although I'll confess that I am not sure why there is a distinction between `β` and `γ`).  It would be much nicer to factor out this code in a helper method located in class EclipseTest.  Something like:
```
class EclipseTest : public testing::Test {
  ...
  void CheckEclipseTime(Instant const& t) {
    auto const q_sun = ...;
    ...
  }
  ...
};
```
and then:
```
CheckEclipseTime(U1);
CheckEclipseTime(U2);
...
```